clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/DefaultIngestionBackend.java

```diff
@@ -50,13 +66,61 @@ public void startConsumption(VeniceStoreVersionConfig storeConfig, int partition
       }
       return storageEngineAtomicReference;
     });
-    LOGGER.info(
-        "Retrieved storage engine for store {} partition {}. Starting consumption in ingestion service",
-        storeVersion,
-        partition);
-    getStoreIngestionService().startConsumption(storeConfig, partition);
+
+    CompletionStage<InputStream> bootstrapFuture =
+        bootstrapFromBlobs(storeAndVersion.getFirst(), storeAndVersion.getSecond().getNumber(), partition);
+
+    bootstrapFuture.whenComplete((result, throwable) -> {
+      LOGGER.info(
+          "Retrieved storage engine for store {} partition {}. Starting consumption in ingestion service",
+          storeVersion,
+          partition);
+      getStoreIngestionService().startConsumption(storeConfig, partition);
+      LOGGER.info(
+          "Completed starting consumption in ingestion service for store {} partition {}",
+          storeVersion,
+          partition);
+    });
+  }
+
+  /**
+   * Bootstrap from the blobs from another source (like another peer). If it fails (due to the 30-minute timeout or
+   * any exceptions), it deletes the partially downloaded blobs, and eventually falls back to bootstrapping from Kafka.
+   * Blob transfer should be enabled to boostrap from blobs, and it currently only supports batch-stores.
+   */
+  private CompletionStage<InputStream> bootstrapFromBlobs(Store store, int versionNumber, int partitionId) {
+    if (!store.isBlobTransferEnabled() || store.isHybrid() || blobTransferManager == null) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    String storeName = store.getName();
+    String baseDir = configLoader.getVeniceServerConfig().getRocksDBPath();
+    CompletionStage<InputStream> p2pTransfer = blobTransferManager.get(storeName, versionNumber, partitionId);
+    CompletableFuture<InputStream> p2pFuture = p2pTransfer.toCompletableFuture();
+
     LOGGER
-        .info("Completed starting consumption in ingestion service for store {} partition {}", storeVersion, partition);
+        .info("Bootstrapping from blobs for store {}, version {}, partition {}", storeName, versionNumber, partitionId);
+
+    CompletableFuture.runAsync(() -> {
+      try {
+        p2pFuture.get(30, TimeUnit.MINUTES);
```

Comment:
I think your implementation is on the right direction but it's a bit overcomplicated and not easy to follow.

I think we can do the followings:
```
// do something
CompletableFuture<InputStream> p2pFuture = p2pTransfer.toCompletableFuture();
// do something
return CompletableFuture.runAsync(() -> {
    try {
            p2pFuture.get(30, TimeUnit.MINUTES);
    } catch (Exception e) {
       // just delete. 
       RocksDBUtils.deleteBlobFiles(...)
      p2pFuture.cancel(true);
      // TODO: we also need to close the channel but let's do it later
    }
}

```

Reply:
Thanks!